bluegreen_deploy.py

```diff
@@ -65,228 +67,328 @@ def list_marathon_apps(args):
-    counts = {}
-    for backend in backends:
-        svname = backend[hmap['svname']]
-        if svname in counts:
-            counts[svname] += 1
-        else:
-            counts[svname] = 1
-        # Are all backends across all instances draining?
-        if counts[svname] == haproxy_instance_count:
-            m = regex.match(svname)
-            host = '.'.join(m.group(1, 2, 3, 4))
-            port = m.group(5)
-            if host in hostports:
-                hostports[host].append(int(port))
-            else:
-                hostports[host] = [int(port)]
-    return hostports
-
-
-def find_tasks_to_kill(tasks, hostports):
-    tasks_to_kill = set()
+def _get_alias_records(hostname):
+    """Return all IPv4 A records for a given hostname
+    """
+    return socket.gethostbyname_ex(hostname)[2]
+
+
+def _unparse_url_alias(url, addr):
+    """Reassemble a url object into a string but with a new address
+    """
+    return urllib.parse.urlunparse((url[0],
+                                    addr + ":" + str(url.port),
+                                    url[2],
+                                    url[3],
+                                    url[4],
+                                    url[5]))
+
+
+def get_marathon_lb_urls(args):
+    """Return a list of urls for all Aliases of the
+       marathon_lb url passed in as an argument
+    """
+    url = urllib.parse.urlparse(args.marathon_lb)
+    addrs = _get_alias_records(url.hostname)
+    return [ _unparse_url_alias(url, addr) for addr in addrs ]
+
+
+def fetch_haproxy_pids(haproxy_url):
+    try:
+        response = requests.get(haproxy_url + "/_haproxy_getpids")
+        response.raise_for_status()
+    except requests.exceptions.RequestException as e:
+        logger.exception("Caught exception when retrieving HAProxy"
+                         " pids from " + haproxy_url)
+        raise e
+
+    return response.text.split()
+
+
+def check_haproxy_reloading(haproxy_url):
+    """Return False if haproxy has only one pid, it is not reloading.
+       Return True if we catch an exception while making a request to
+       haproxy or if more than one pid is returned
+    """
+    try:
+        pids = fetch_haproxy_pids(haproxy_url)
+    except requests.exceptions.RequestException as e:
+        return True
+
+    if len(pids) > 1:
+        logger.info("Waiting for {} pids on {}".format(len(pids), haproxy_url))
+        return True
+
+    return False
+
+
+def any_marathon_lb_reloading(marathon_lb_urls):
+    return any([check_haproxy_reloading(url) for url in marathon_lb_urls])
+
+
+def fetch_haproxy_stats(haproxy_url):
+    try:
+        response = requests.get(haproxy_url + "/haproxy?stats;csv")
+        response.raise_for_status()
+    except requests.exceptions.RequestException as e:
+        logger.exception("Caught exception when retrieving HAProxy"
+                         " stats from " + haproxy_url)
+        raise e
```

Comment:
Don't need `raise e` here. Just `raise` will reraise the current exception (and, in Python 2, will preserve the stack trace)